src/Microsoft.Data.Analysis/DataFrame.Join.cs

```diff
@@ -155,207 +164,220 @@ public DataFrame Join(DataFrame other, string leftSuffix = "_left", string right
-                            rightRowIndices.Append(i);
-                        }
-                    }
-                    else
+                    //only rows with occurences from previose step should go for futher processing
+                    var shrinkedRetainedIndices = occurrences.Keys.ToArray();
+
+                    //create reverse mapping of index of the row in the shrinked column to the index of this row in the original dataframe (new index -> original index)
+                    var newRetainedIndicesReverseMapping = new Dictionary<long, long>(shrinkedRetainedIndices.Length);
+
+                    for (int i = 0; i < shrinkedRetainedIndices.Length; i++)
                     {
-                        foreach (long thisColumnNullIndex in thisColumnNullIndices)
-                        {
-                            leftRowIndices.Append(thisColumnNullIndex);
-                            rightRowIndices.Append(i);
-                        }
+                        //store reverse mapping to restore original dataframe indices from indices in shrinked row
+                        var originalIndex = shrinkedRetainedIndices[i];
+                        newRetainedIndicesReverseMapping.Add(i, originalIndex);
                     }
+
+                    retainedIndicesReverseMapping = newRetainedIndicesReverseMapping;
+                    shrinkedRetainedColumn = shrinkedRetainedColumn.Clone(new Int64DataFrameColumn("Indices", shrinkedRetainedIndices));
                 }
-            }
-            else if (joinAlgorithm == JoinAlgorithm.Inner)
-            {
-                // Hash the column with the smaller RowCount 
-                long leftRowCount = Rows.Count;
-                long rightRowCount = other.Rows.Count;
+                
+                DataFrameColumn supplementaryColumn = supplementaryDataFrame.Columns[supplemetaryJoinColumnNames[colNameIndex]];
+
+                //Find occurrenses on current step (join column)
+                var newOccurrences = shrinkedRetainedColumn.GetGroupedOccurrences(supplementaryColumn, out HashSet<long> supplementaryColumnNullIndices);
 
-                bool leftColumnIsSmaller = leftRowCount <= rightRowCount;
-                DataFrameColumn hashColumn = leftColumnIsSmaller ? Columns[leftJoinColumn] : other.Columns[rightJoinColumn];
-                DataFrameColumn otherColumn = ReferenceEquals(hashColumn, Columns[leftJoinColumn]) ? other.Columns[rightJoinColumn] : Columns[leftJoinColumn];
-                Dictionary<TKey, ICollection<long>> multimap = hashColumn.GroupColumnValues<TKey>(out HashSet<long> smallerDataFrameColumnNullIndices);
+                //Convert indices from in key from local (shrinked row) to indices in original dataframe
+                if (retainedIndicesReverseMapping != null)
+                    newOccurrences = newOccurrences.ToDictionary(kvp => retainedIndicesReverseMapping[kvp.Key], kvp => kvp.Value);
```

Comment:
Going from array->dictionary here. Another place to consider performance.

Reply:
we are going from dictionary to dictionary, the idea is that we have to convert indices from _shrinkedRetainedColumn_ (to wich we apply GetGroupedOccurrences) to the indices for original data frame. Possible solution is to provide _retainedIndicesReverseMapping_ inside 
_protected Dictionary<long, ICollection<long>> GetGroupedOccurrences<TKey>(DataFrameColumn other, out HashSet<long> otherColumnNullIndices)_ as optional parameter to create correct indices just in place. What do you think, should we do that?

Reply:
Yup, you're right, it's dictionary -> dictionary. The diff on GH made it hard to find out the type of newOccurences. I think this is ok. It's more readable IMO